File: Operators/search_eda.m
function [output1,output2] = search_eda(varargin)
% The estimation of distribution.

mode = varargin{end};
switch mode
    case 'execute'
        Parent = varargin{1};

        Parent = Parent.decs;
        [N,D]  = size(Parent);
        
        output1 = zeros(N,D);
        for i = 1:D
            pd = fitdist(Parent(:,i),'normal'); % fit a normal distribution
            output1(:,i) = random(pd,[N,1]); % sample for the fitted distribution
        end
        output2 = varargin{5};
  
    case 'parameter'
        % n/a

    case 'behavior'
        output1 = {'';'GS'}; % always performs global search
end

if ~exist('output1','var')
    output1 = [];
end
if ~exist('output2','var')
    output2 = [];
end
end

File: src/autooptlib/components/search_eda.py
"""Estimation of Distribution Algorithm (EDA) operator."""

from __future__ import annotations

from typing import Any

import numpy as np


def search_eda(*args: Any):
    mode = args[-1]
    if mode == "execute":
        parent = args[0]
        aux = args[3] if len(args) > 3 else None

        decs = getattr(parent, "decs", None)
        if callable(decs):
            parent = decs()
        else:
            parent = decs if decs is not None else parent
        parent = np.asarray(parent, dtype=float)
        if parent.ndim != 2:
            raise ValueError("Parent decision variables must form a 2D array")
        n, _ = parent.shape
        mean = parent.mean(axis=0)
        std = parent.std(axis=0, ddof=1)
        std = np.where(np.isfinite(std) & (std > 0), std, parent.std(axis=0, ddof=0))
        std = np.where(std > 0, std, 1e-12)
        offspring = np.random.normal(loc=mean, scale=std, size=(n, parent.shape[1]))
        return offspring, aux

    if mode == "parameter":
        return None, None

    if mode == "behavior":
        return ["", "GS"], None

    raise ValueError(f"Unsupported mode: {mode}")

